hypershift-operator/controllers/hostedcluster/internal/platform/agent/agent.go

```diff
@@ -0,0 +1,88 @@
+package agent
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	agentv1 "github.com/eranco74/cluster-api-provider-agent/api/v1alpha1"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests/controlplaneoperator"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests/ignitionserver"
+	"github.com/openshift/hypershift/support/upsert"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	capiv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Agent struct{}
+
+func (p Agent) ReconcileCAPIInfraCR(ctx context.Context, c client.Client, createOrUpdate upsert.CreateOrUpdateFN,
+	hcluster *hyperv1.HostedCluster,
+	controlPlaneNamespace string, apiEndpoint hyperv1.APIEndpoint) (client.Object, error) {
+
+	hcp := controlplaneoperator.HostedControlPlane(controlPlaneNamespace, hcluster.Name)
+	if err := c.Get(ctx, client.ObjectKeyFromObject(hcp), hcp); err != nil {
+		return nil, fmt.Errorf("failed to get control plane ref: %w", err)
+	}
+
+	caSecret := ignitionserver.IgnitionCACertSecret(hcp.Namespace)
+	encodedCACert := ""
+	if err := c.Get(ctx, client.ObjectKeyFromObject(caSecret), caSecret); err == nil {
+		caCertBytes, hasCACert := caSecret.Data[corev1.TLSCertKey]
+		if !hasCACert {
+			return nil, fmt.Errorf("CA Secret is missing tls.crt key")
+		}
+		encodedCACert = base64.StdEncoding.EncodeToString(caCertBytes)
+	}
+
+	agentCluster := &agentv1.AgentCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: controlPlaneNamespace,
+			Name:      hcluster.Name,
+		},
+	}
+
+	_, err := createOrUpdate(ctx, c, agentCluster, func() error {
+		return reconcileAgentCluster(agentCluster, hcluster, hcp, encodedCACert)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return agentCluster, nil
+}
+
+func (p Agent) CAPIProviderDeploymentSpec(hcluster *hyperv1.HostedCluster, tokenMinterImage string) (*appsv1.DeploymentSpec, error) {
+	return nil, nil
```

Comment:
Why there's no controller to watch agent cluster/machines?

Reply:
Our CAPI provider watches them - it needs to be deployed separately.  There is one instance for all hostedclusters.

Reply:
>it needs to be deployed separately

who/how/when is that deployed?

Reply:
Whoever deploys Assisted Installer itself is expected to also deploy the CAPI provider.
We can consider having hypershift deploy the operator per namespace (it should work in theory) but there will still be one instance of Assisted with non-namespaced CRs.

Reply:
Is that whoever also responsible for installing the CAPI platform CRDs? e.g in AWS https://github.com/openshift/hypershift/tree/main/cmd/install/assets/cluster-api-provider-aws

Reply:
Hmm.
I think (as @avishayt said) that the capi-provider-agent gets installed together with the assisted-installer (it should probably come as an option in the infrastructure operator)
Unlike capi-provider-aws that is installed by hypershift (and just works when hypershift run on AWS) the provider agent  is an integral part of the assisted-installer and hypershift can assume its already installed.

Reply:
ok I'm good proceeding with this being an out of band req managed by assisted installer, we can iterate later if we need to.
Please add a comment explaining this context to your `CAPIProviderDeploymentSpec`  implementation.